packages/nx/src/project-graph/build-dependencies/extract-file-references.ts

```diff
@@ -0,0 +1,488 @@
+                break;
+              case SyntaxKind.OpenBraceToken:
+                if (stack.length) {
+                  stack.push(token);
+                }
+                break;
+              case SyntaxKind.CloseBraceToken:
+                if (stack.length) {
+                  if (lastOrUndefined(stack) === SyntaxKind.TemplateHead) {
+                    if (
+                      scanner.reScanTemplateToken(
+                        /* isTaggedTemplate */ false
+                      ) === SyntaxKind.TemplateTail
+                    ) {
+                      stack.pop();
+                    }
+                  } else {
+                    stack.pop();
+                  }
+                }
+                break;
+            }
+            token = scanner.scan();
+          }
+          nextToken();
+        }
+
+        // Check if at least one of the alternatives has moved the scanner forwards
+        if (
+          tryConsumeImport() ||
+          tryConsumeExport() ||
+          tryConsumeRequireCall(
+            /*skipCurrentToken*/ false,
+            /*allowTemplateLiterals*/ true
+          )
+        ) {
+          continue;
+        } else {
+          nextToken();
+        }
+      }
+
+      scanner.setText(undefined);
+    }
+
+    processImports();
+
+    /**
+     * Extract any <reference path="..."> pragmas from the file
+     */
+    for (const range of ts.getLeadingCommentRanges(sourceText, 0) || []) {
+      const comment = sourceText.substring(range.pos, range.end);
+      const tripleSlash =
+        range.kind === SyntaxKind.SingleLineCommentTrivia &&
+        tripleSlashXMLCommentStartRegEx.exec(comment);
+      if (!tripleSlash) {
+        continue;
+      }
+      const pragmaName = tripleSlash[1].toLowerCase();
+      if (pragmaName !== 'reference') {
+        continue;
+      }
+      const matchResult = referencePathArgRegEx.exec(comment);
+      if (!matchResult) {
+        continue; // The <reference /> does not have a path argument, ignore it
+      }
+      const value = matchResult[2] || matchResult[3];
+      const startPos =
+        range.pos + matchResult.index + matchResult[1].length + 1;
+
+      allFileReferences.push({
+        pos: startPos,
+        end: startPos + value.length,
+        fileName: value,
+      });
+    }
+
+    /**
+     * We only need to build a full ts.SourceFile (AST) if at least one of the following is true:
+     *
+     * 1. The source contains one or more nx-ignore-next-line comments
+     * 2. The source contains legacy Angular loadChildren string syntax
+     */
+    let sourceFile: SourceFile | undefined;
+    let lineStarts: number[] | undefined;
+
+    const ignoredLines = new Set();
+
+    const totalNxIgnoreNextLineOccurrences =
+      sourceText.split(NX_IGNORE_NEXT_LINE_COMMENT).length - 1;
+    if (totalNxIgnoreNextLineOccurrences > 0) {
+      if (!sourceFile) {
+        sourceFile = getSourceFile(filePath, sourceText);
```

Comment:
Using the AST is just much more reliable that's all - as you can see in the diff I had already identified some broken cases for `nx-ignore-next-line` comments not being interpreted correctly which are fixed by this PR, and there were possibly others because I didn't spend too much time on it.

Given `nx-ignore-next-line` is so rarely used (and currently undocumented I think), it seemed like a good balance to be super fast with only the scanner in the common case, but then being _accurate_ in the uncommon case of `nx-ignore-next-line` being leveraged in a file